File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.py
# 전화번호 목록

def solution(phone_book):
    dict = {}

    # 각 전화번호의 해시 값을 계산하여 해시 맵에 저장(여기 저장하는게 중요)
    for i in phone_book:
        dict[i] = True

    # 각 전화번호의 접두어가 해시 맵에 있는지 확인
    for i in phone_book:
        prefix = ""
        for digit in i:
            # prefix라는 변수에 digit을 하나씩 추가하여 접두어 형성         
            prefix += digit      
            
            # 접두어가 해시 맵에 있다면 해당 전화번호는 다른 번호의 접두어
            # 자기 자신을 접두어로 판단하지 않기 위해 i는 빼고 체크
            if prefix != i and prefix in dict:
                return False
    
    return True

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.java
import java.util.Arrays;

class Solution {
    // 한 번호가 다른 번호의 접두어인 경우 false
    // 아니면 true
    public boolean solution(String[] phone_book) {
        // 1. 전화번호부 정렬
        Arrays.sort(phone_book);

        // 2. 전화번호부에서 연속된 두 개의 전화번호 비교
        for (int i = 0; i < phone_book.length - 1; i++) {
            // phone_book[i+1] 문자열이 phone_book[i] 문자열로 시작하는지를 확인
            if (phone_book[i + 1].startsWith(phone_book[i]))
                return false;
        }
        return true;
    }
}